main-command/src/main/scala/sbt/internal/client/NetworkClient.scala

```diff
@@ -9,94 +9,242 @@ package sbt
+      useJNI = false,
+    )
   private val status = new AtomicReference("Ready")
   private val lock: AnyRef = new AnyRef {}
   private val running = new AtomicBoolean(true)
-  private val pendingExecIds = ListBuffer.empty[String]
+  private val pendingResults = new ConcurrentHashMap[String, (LinkedBlockingQueue[Integer], Long)]
+  private val pendingCancellations = new ConcurrentHashMap[String, LinkedBlockingQueue[Boolean]]
+  private val pendingCompletions = new ConcurrentHashMap[String, CompletionResponse => Unit]
+  private val attached = new AtomicBoolean(false)
+  private val attachUUID = new AtomicReference[String](null)
+  private val connectionHolder = new AtomicReference[ServerConnection]
+  private val batchMode = new AtomicBoolean(false)
+  private val interactiveThread = new AtomicReference[Thread](null)
+  private lazy val noTab = arguments.completionArguments.contains("--no-tab")
+  private lazy val noStdErr = arguments.completionArguments.contains("--no-stderr") &&
+    System.getenv("SBTC_AUTO_COMPLETE") == null
+
+  private def mkSocket(file: File): (Socket, Option[String]) = ClientSocket.socket(file, useJNI)
 
-  private val console = ConsoleAppender("thin1")
-  private def baseDirectory: File = configuration.baseDirectory
+  private def portfile = arguments.baseDirectory / "project" / "target" / "active.json"
 
-  lazy val connection = init()
+  def connection: ServerConnection = connectionHolder.synchronized {
+    connectionHolder.get match {
+      case null => init(prompt = false, retry = true)
+      case c    => c
+    }
+  }
 
-  start()
+  private[this] val stdinBytes = new LinkedBlockingQueue[Int]
+  private[this] val stdin: InputStream = new InputStream {
+    override def available(): Int = stdinBytes.size
+    override def read: Int = stdinBytes.take
+  }
+  private[this] val inputThread = new AtomicReference(new RawInputThread)
+  private[this] val exitClean = new AtomicBoolean(true)
+  private[this] val sbtProcess = new AtomicReference[Process](null)
+  private class ConnectionRefusedException(t: Throwable) extends Throwable(t)
 
   // Open server connection based on the portfile
-  def init(): ServerConnection = {
-    val portfile = baseDirectory / "project" / "target" / "active.json"
-    if (!portfile.exists) {
-      forkServer(portfile)
-    }
-    val (sk, tkn) = ClientSocket.socket(portfile)
-    val conn = new ServerConnection(sk) {
-      override def onNotification(msg: JsonRpcNotificationMessage): Unit = self.onNotification(msg)
-      override def onRequest(msg: JsonRpcRequestMessage): Unit = self.onRequest(msg)
-      override def onResponse(msg: JsonRpcResponseMessage): Unit = self.onResponse(msg)
-      override def onShutdown(): Unit = {
-        running.set(false)
+  def init(prompt: Boolean, retry: Boolean): ServerConnection =
+    try {
+      if (!portfile.exists) {
+        if (prompt) {
+          val msg = if (noTab) "" else "No sbt server is running. Press <tab> to start one..."
```

Comment:
Btw, I instinctively hit 'y' for this, today.  But I'm not sure why.  We use 'r' for retrying to (re)load, so where is my muscle memory remembering 'y' from?

Just because from that I'm wondering if it should be allowed.  "Press tab to confirm" feels a little unique/original here.  (No big deal.)

Reply:
I have a vague memory of tab completion doing something special on double tabs.
From `prompt` it's not obvious, but this `prompt` only becomes true during the completion service call right? (Maybe rename the variable to reflect that?)

Reply:
For example in zsh

```
$ ls g[tab]
```

completes to

```
$ ls giter8-bootstrap_2.12-0.13.
```

and hitting the second tab

```
$ ls giter8-bootstrap_2.12-0.13.[tab]
```

completes to

```
$ ls giter8-bootstrap_2.12-0.13.
giter8-bootstrap_2.12-0.13.0.sh  giter8-bootstrap_2.12-0.13.1.sh
```

and hitting the third tab

```
$ ls giter8-bootstrap_2.12-0.13.[tab]
giter8-bootstrap_2.12-0.13.0.sh  giter8-bootstrap_2.12-0.13.1.sh
```

completes to

```
$ ls giter8-bootstrap_2.12-0.13.0.sh
giter8-bootstrap_2.12-0.13.0.sh  giter8-bootstrap_2.12-0.13.1.sh
```

Reply:
The motivation for using <tab> that if I know that an sbt server isn't running, I can just double tap tab. Also, as Eugene pointed out, it's not that uncommon for pressing tab multiple times to modify the completions. I'm not opposed to adding 'y' as another valid option. The main thing is that I wanted an opt out in case users don't want tab completions to do something potentially slow.

Reply:
Yeah, I agree to some way to confirm first.  But I feel that double-tabbing is normally used to avoid filling a user's screen, rather than performance reasons.  But that's not to say I'm against using tab for it.

I'm happy to leave things as is, until I can figure out what might've prompted my instinct.

Reply:
Yeah. That's true that double tab is usually used to avoid spamming the screen. I feel like this is nevertheless similar given that in both situations the program delays displaying the completions until the user presses tab a second time.

Reply:
https://github.com/sbt/sbt/pull/5620/commits/44a98186b707dccc3018fac95dcaef76b237a834